File: src/main/java/com/tomatokey/prototype/configuration/datasource/DynamicRoutingDataSource.java
package com.tomatokey.prototype.configuration.datasource;

import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource;

import java.util.Objects;

public class DynamicRoutingDataSource extends AbstractRoutingDataSource {

    @Override
    protected Object determineCurrentLookupKey() {
        if (Objects.isNull(MultiDataSourceContextHolder.getDataSourceType())) {
            return DataSourceType.REF;
        }
        return MultiDataSourceContextHolder.getDataSourceType();
    }

}

File: src/main/kotlin/com/prototype/framework/configuration/jdbc/datasource/DynamicRoutingDataSource.kt
package com.prototype.framework.configuration.jdbc.datasource

import org.springframework.jdbc.datasource.lookup.AbstractRoutingDataSource

class DynamicRoutingDataSource : AbstractRoutingDataSource() {
    override fun determineCurrentLookupKey(): Any? {
        return MultiDataSourceContextHolder.dataSourceType ?: resolvedDefaultDataSource
    }
}